fn/async.go

```diff
@@ -0,0 +1,69 @@
+package fn
+
+import "errors"
+
+type Async[A any] interface {
+	Await() (*A, error)
+	Cancel()
+}
+
+// RunAsyncWithCancel is a function that will transform a synchronous function
+// and make it a cancellable asynchronous call. This can be useful if the
+// function in the argument can block indefinitely and the calling context needs
+// a way to abandon waiting for it to complete.
+func RunAsyncWithCancel[A any](
+	f func() (*A, error),
+	cancelSig chan struct{},
+	cancelErr error,
+) Async[A] {
+	a := async[A]{
+		quit: cancelSig,
+		res: make(chan *A),
+		err: make(chan error),
+		cancelErr: cancelErr,
+	}
+	go func() {
+		res, err := f()
+		if err != nil {
+			select {
+			case a.err <- err:
+			case <-a.quit:
+			}
+		} else {
+			select {
+			case a.res <- res:
+			case <-a.quit:
+			}
+		}
+	}()
+
+	return &a
+}
+
+func RunAsync[A any](f func() (*A, error)) Async[A] {
+	return RunAsyncWithCancel(
+		f, make(chan struct{}), errors.New("async cancelled"),
+	)
+}
+
+func (a *async[A]) Await() (*A, error) {
+	select {
+	case r := <-a.res:
+		return r, nil
+	case e := <-a.err:
+		return nil, e
+	case <-a.quit:
+		return nil, a.cancelErr
+	}
+}
+
+func (a *async[A]) Cancel() {
+	close(a.quit)
```

Comment:
This is the main thing I'm concerned about. If we tie the Async action to a larger context then the Cancel function will cancel the whole context. So it's possible that `RunAsyncWithCancel` is an unsafe API that we don't want to expose.

Reply:
With the current API, isn't `async.quit` always a newly created chan?  How exactly would `quit` get tied to some other chan?

Reply:
Probably we want two chans -- one to cancel from outside, and one to cancel by calling `Cancel`.

Reply:
🤔